File: path_sum_ii/path_sum_ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    # This is not really the most optimal solution as I'm looping through every node in the tree structure.
    # and within that loop I'm copying the path state and then checking if the sum lines up with targetSum.
    # The time complexity of this solution at there for me regarded as O(n^2) where n is the number of nodes in the tree structure. This is because for every node in the tree structure we are copying the state of the path and then checking if the sum lines up with targetSum.
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        # okay for this problem I will need to have a results array that will keep track of all the possible paths that lead to the target sum as well as the state of the current recursive loop.

        results = []

        self.recursive_helper(
            [],
            root,
            targetSum, 
            results
        )

        return results

    def recursive_helper(self, state: List[int], treeNode: Optional[TreeNode], targetSum: int, results: List[List[int]]):
        # create a conditional that will check if the sum of state equals the target sum
        # if it does than we can add the copy of the state to the results array

        # create a conditional that check if the current node is a treeNode or null
        # if a treeNode 
        # add current node value to state
        # recursively call the function on the right and left nodes

        if treeNode != None and treeNode.left == None and treeNode.right == None:
            result_state = [*state, treeNode.val]
            if len(result_state) != 0 and sum(result_state) == targetSum:
                results.append(result_state)

        if treeNode != None:
            if treeNode.left:
                self.recursive_helper([*state, treeNode.val], treeNode.left, targetSum, results)
            if treeNode.right:
                self.recursive_helper([*state, treeNode.val], treeNode.right, targetSum, results)

File: path_sum_ii/path_sum_ii.go
package pathsumii

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// The time complexity for this solution is more in line with O(n) since unlike the python solution I'm
// not copying the state of the path for every recursive call.
func pathSum(root *TreeNode, targetSum int) [][]int {
	// first create an array that will carry the resulting paths of the depth first search traversal.

	// create a recursive function that will be used to loop through the entire tree structure and add the node to leaf paths that add up to the target sum.

	// return results

	results := [][]int{}

	recursiveHelper(
		[]int{},
		root,
		targetSum,
		&results,
	)

	return results
}

// recursiveHelper function will take in state, treeNode, targetSum, and results. and will not return anything since it will be writing to a pointer of results
func recursiveHelper(state []int, treeNode *TreeNode, targetSum int, results *[][]int) {
	if treeNode != nil && treeNode.Left == nil && treeNode.Right == nil {
		if (targetSum - treeNode.Val) == 0 {
			dupState := append([]int{}, state...)
			dupState = append(dupState, treeNode.Val)
			*results = append(*results, dupState)
		}
		return
	}

	if treeNode != nil {
		state = append(state, treeNode.Val)
		if treeNode.Left != nil {
			recursiveHelper(
				state,
				treeNode.Left,
				targetSum-treeNode.Val,
				results,
			)
		}

		if treeNode.Right != nil {
			recursiveHelper(
				state,
				treeNode.Right,
				targetSum-treeNode.Val,
				results,
			)
		}
	}
}
